### DISCUSSION TITLE
Creating custom packages guidelines

### DISCUSSION THREAD

**Discussion by spowell-216** [iTwin/itwinjs-core#1882] [Q&A] [answered]
Hello,
I am trying to segment out some iModeljs code that works in one project and package for use across multiple projects but running into issues when executing, specifically when trying to create models.  These projects are all backend projects so I'm only using tsc 1>&2 to the build the package.

currently both project and package are on the same iModeljs ver: 2.16.1

The error i keep getting when trying to create definition model: 

"TypeError: Cannot read property 'schemaName' of undefined."

Are there any special build steps i might be missing?

here's the two snippets i have tried and were successful in a project as well as unit testing against a Snapshotdb.

```
// Get/Create Subject
const parentSubjectId = targetIModelDb.elements.getRootSubject().id;
// Get/Create Definition Model
let targetDefinitionModelId = targetIModelDb.elements.queryElementIdByCode(DefinitionPartition.createCode(targetIModelDb, parentSubjectId, "definitions"));
if (targetDefinitionModelId == undefined){
  targetDefinitionModelId = DefinitionModel.insert(targetIModelDb, parentSubjectId, "definitions");
}

```
```
// Get/Create Subject
const parentSubjectId = targetIModelDb.elements.getRootSubject().id;

// Get/Create Definition Model
let targetDefinitionModelId = targetIModelDb.elements.queryElementIdByCode(DefinitionPartition.createCode(targetIModelDb, parentSubjectId, "definitions"));
if (targetDefinitionModelId == undefined){
  Logger.logTrace(this.loggingCategory, `DefinitionPartition not found, creating a new one`);
  const defModelProps: ElementProps = {
    classFullName: DefinitionPartition.classFullName,
    model: IModel.repositoryModelId,
    parent: new SubjectOwnsPartitionElements(IModel.rootSubjectId),
    code: DefinitionPartition.createCode(targetIModelDb, IModel.rootSubjectId, "definitions")
  }
  const partitionId = targetIModelDb.elements.insertElement(defModelProps);
  targetDefinitionModelId = targetIModelDb.models.insertModel({classFullName: DefinitionModel.classFullName, modeledElement: { id: partitionId }});
}
```
This package is successful in 
1. downloading a new iModel from Hub
2. importing XML schema file
3. registering schema and classes
but fails when trying to insert the models.

**Comment by pmconne** [comment]
Have you tried using the debugger to obtain a stack trace and inspect which variable is undefined?

**Comment by spowell-216** [comment, marked as answer]
yes, sorry forgot to include it in the initial post
![image](https://user-images.githubusercontent.com/51784707/125980694-6a975e77-9a25-4bb6-8a06-343ac7a32c97.png)

**Reply by aruniverse** [reply]
You mentioned this is going to be a backend package. Curious how you have the dependencies set up in your backend package's package.json. I'd recommend using dev+peer deps to make sure there is only one instance of imodeljs-backend and other imodeljs packages in your app. If it's defined as a direct dependency you might have more than 1  copy of those singletons in your app causing type mismatches and doppelganger issues with certain objects being undefined.

**Reply by spowell-216** [reply]
Thank you, this looks to be what i am overlooking.  I will give it a try and get back if i am still running into the same problem.

**Reply by pmconne** [reply]
Please mark the question as answered if the problem is resolved.
